### DISCUSSION TITLE
WebSocket disconnect when idleTimeout is less than 6 seconds

### DISCUSSION THREAD

**Discussion by DarkMatterMatt** [uNetworking/uWebSockets.js#286] [Q&A] [answered]
Hey, I've found that websockets will be dropped when idleTimeout is less than 6 seconds.

**Method of testing:**
Run a websocket with a short timeout
Go to http://crysislinux.github.io/smart_websocket_client/
Connect to the websocket and spam a message (connection should remain until the server has not received a message for 5 seconds)
Watch as it drops the connection.

**Sample code**, using uWS v17.3.0

    require('uWebSockets.js').App().ws('/*', {
        idleTimeout: 5.9999, // disconnects
        // idleTimeout: 6, // doesn't disconnect

        message: (ws, message, isBinary) => {
            let ok = ws.send(message, isBinary);
        }
    }).listen(9001, (listenSocket) => {
        if (listenSocket) {
            console.log('Listening to port 9001');
        }
    });

**Comment by ghost** [comment]
What if you specify 6.9999

**Comment by DarkMatterMatt** [comment]
| Timeout | Disconnects |
|---------|-------------|
| 4       | ✓           |
| 5       | ✓           |
| 5.5     | ✓           |
| 5.9999  | ✓           |
| 6       | ✗           |
| 6.5     | ✗           |
| 6.9999  | ✗           |
| 7       | ✗           |

Edit to add screenshot (`idleTimeout = 5`):
![uWebSocket disconnects](https://user-images.githubusercontent.com/15150339/77210424-d9e8c100-6b65-11ea-9a0c-4286a27b5f89.png)

    x = new WebSocket("ws://localhost:9001/");
    x.addEventListener("open", () => console.log("open ", (new Date()).getTime()));
    x.addEventListener("close", ev => console.log("close", (new Date()).getTime(), ev.code, ev.reason));

**Comment by ghost** [comment, marked as answer]
I cannot reproduce this issue. Your screenshot seems to show how it is supposed to act. Timeout is very granular and works in steps of 4 seconds. This means 5 seconds will round to 4 seconds which is anything between 0 to 4 seconds if I remember correctly. This while 6 will round to 8 seconds, which is anything between 4 to 8 seconds.

This is controlled by LIBUS_TIMEOUT_GRANULARITY which is 4 seconds by default.

**Comment by DarkMatterMatt** [comment]
Interesting, I didn't know about `LIBUS_TIMEOUT_GRANULARITY`, thanks :)
